File: mage/src/main/java/mil/nga/giat/mage/login/SignupActivity.java
package mil.nga.giat.mage.login;

import android.content.Intent;
import android.content.SharedPreferences;
import android.graphics.Typeface;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.Patterns;
import android.view.KeyEvent;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.textfield.TextInputLayout;

import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import mil.nga.giat.mage.R;
import mil.nga.giat.mage.sdk.login.SignupStatus;
import mil.nga.giat.mage.sdk.login.SignupTask;
import mil.nga.giat.mage.sdk.preferences.PreferenceHelper;

public class SignupActivity extends AppCompatActivity implements SignupTask.SignupDelegate {
	private static final String LOG_NAME = SignupActivity.class.getName();

	private EditText mDisplayNameEditText;
	private TextInputLayout mDisplayNameLayout;

	private EditText mUsernameEditText;
	private TextInputLayout mUsernameLayout;

	private EditText mEmailEditText;
	private TextInputLayout mEmailLayout;

	private EditText mPhoneEditText;

	private EditText mPasswordEditText;
	private TextInputLayout mPasswordLayout;

	private EditText mConfirmPasswordEditText;
	private TextInputLayout mConfirmPasswordLayout;

	private String serverURL;

	public final EditText getDisplayNameEditText() {
		return mDisplayNameEditText;
	}

	public final EditText getUsernameEditText() {
		return mUsernameEditText;
	}

	public final EditText getEmailEditText() {
		return mEmailEditText;
	}

	public final EditText getPhoneEditText() {
		return mPhoneEditText;
	}

	public final EditText getPasswordEditText() {
		return mPasswordEditText;
	}

	public final EditText getConfirmPasswordEditText() {
		return mConfirmPasswordEditText;
	}

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		setContentView(R.layout.activity_signup);

		TextView appName = findViewById(R.id.mage);
		appName.setTypeface(Typeface.createFromAsset(getAssets(),"fonts/GondolaMage-Regular.otf"));

		final PasswordStrengthFragment passwordStrengthFragment = (PasswordStrengthFragment) getSupportFragmentManager().findFragmentById(R.id.password_strength_fragment);

		mDisplayNameEditText = findViewById(R.id.signup_displayname);
		mDisplayNameLayout = findViewById(R.id.displayname_layout);

		mUsernameEditText = findViewById(R.id.signup_username);
		mUsernameLayout = findViewById(R.id.username_layout);

		mEmailEditText = findViewById(R.id.signup_email);
		mEmailLayout = findViewById(R.id.email_layout);

		mPhoneEditText = findViewById(R.id.signup_phone);

		mPasswordEditText = findViewById(R.id.signup_password);
		mPasswordEditText.setTypeface(Typeface.DEFAULT);
		mPasswordLayout = findViewById(R.id.password_layout);
		mPasswordEditText.addTextChangedListener(new TextWatcher() {
			@Override
			public void beforeTextChanged(CharSequence s, int start, int count, int after) {

			}

			@Override
			public void onTextChanged(CharSequence s, int start, int before, int count) {

			}

			@Override
			public void afterTextChanged(Editable s) {
				List<String> sanitizedPasswordInputs = new ArrayList<>();
				sanitizedPasswordInputs.add(mDisplayNameEditText.getText().toString());
				sanitizedPasswordInputs.add(mUsernameEditText.getText().toString());
				sanitizedPasswordInputs.add(mEmailEditText.getText().toString());
				sanitizedPasswordInputs.removeAll(Collections.singleton(null));
				passwordStrengthFragment.setSanitizedList(sanitizedPasswordInputs);

				passwordStrengthFragment.onPasswordChanged(s.toString());
			}
		});

		mConfirmPasswordEditText = findViewById(R.id.signup_confirmpassword);
		mConfirmPasswordEditText.setTypeface(Typeface.DEFAULT);
		mConfirmPasswordLayout = findViewById(R.id.confirmpassword_layout);

		SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
		serverURL = sharedPreferences.getString(getString(R.string.serverURLKey), getString(R.string.serverURLDefaultValue));

		mConfirmPasswordEditText.setOnKeyListener((v, keyCode, event) -> {
			if (keyCode == KeyEvent.KEYCODE_ENTER) {
				signup(v);
				return true;
			} else {
				return false;
			}
		});

		configureSignup();
	}

	private void configureSignup() {
		PreferenceHelper preferenceHelper = PreferenceHelper.getInstance(getApplicationContext());
		if (preferenceHelper.containsLocalAuthentication()) {
			Button localButton = findViewById(R.id.local_signup_button);
			localButton.setOnClickListener(this::signup);
		}
	}

	/**
	 * Fired when user clicks login
	 *
	 * @param view
	 */
	public void login(View view) {
		startActivity(new Intent(getApplicationContext(), LoginActivity.class));
		finish();
	}


	/**
	 * Fired when user clicks signup
	 *
	 * @param view
	 */
	public void signup(View view) {
		// reset errors
		mDisplayNameLayout.setError(null);
		mUsernameLayout.setError(null);
		mEmailLayout.setError(null);
		mPasswordLayout.setError(null);
		mConfirmPasswordLayout.setError(null);

		String displayName = getDisplayNameEditText().getText().toString();
		String username = getUsernameEditText().getText().toString();
		String email = getEmailEditText().getText().toString();
		String phone = getPhoneEditText().getText().toString();
		String password = getPasswordEditText().getText().toString();
		String confirmpassword = getConfirmPasswordEditText().getText().toString();

		// are the inputs valid?
		if (TextUtils.isEmpty(displayName)) {
			mDisplayNameLayout.setError("Display name can not be blank");
			return;
		}

		if (TextUtils.isEmpty(username)) {
			mUsernameLayout.setError("Username can not be blank");
			return;
		}

		// is email address the right syntax?
		if (StringUtils.isNotEmpty(email) && !Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
			mEmailLayout.setError("Please enter a valid email address");
			return;
		}

		if (TextUtils.isEmpty(password)) {
			mPasswordLayout.setError("Password can not be blank");
			return;
		}

		if (TextUtils.isEmpty(confirmpassword)) {
			mConfirmPasswordLayout.setError("Enter password again");
			return;
		}

		// do passwords match?
		if (!password.equals(confirmpassword)) {
			mPasswordLayout.setError("Passwords do not match");
			mConfirmPasswordLayout.setError("Passwords do not match");
			return;
		}

		List<String> accountInfo = new ArrayList<>();
		accountInfo.add(username);
		accountInfo.add(displayName);
		accountInfo.add(email);
		accountInfo.add(phone);
		accountInfo.add(password);
		accountInfo.add(serverURL);

		// show spinner, and hide form
		findViewById(R.id.signup_form).setVisibility(View.GONE);
		findViewById(R.id.signup_status).setVisibility(View.VISIBLE);

		new SignupTask(this.getApplicationContext(), this).execute(accountInfo.toArray(new String[accountInfo.size()]));
	}

	@Override
	public void onSignupComplete(SignupStatus status) {
		if (status.getStatus() == SignupStatus.Status.SUCCESSFUL_SIGNUP) {
			boolean isActive = status.getUser().get("active").getAsBoolean();

			// Tell the user that their account was made
			showSignupSuccessDialog(isActive);
		} else {
			new AlertDialog.Builder(this)
					.setTitle("Signup Failed")
					.setMessage(status.getMessage())
					.setPositiveButton(android.R.string.ok, null)
					.show();

			// show form, and hide spinner
			findViewById(R.id.signup_status).setVisibility(View.GONE);
			findViewById(R.id.signup_form).setVisibility(View.VISIBLE);
		}
	}

	@Override
	protected void onResume() {
		super.onResume();
		// show form, and hide spinner
		findViewById(R.id.signup_status).setVisibility(View.GONE);
		findViewById(R.id.signup_form).setVisibility(View.VISIBLE);
	}

	private void showSignupSuccessDialog(boolean isActive) {
		AlertDialog.Builder alertDialog = new AlertDialog.Builder(this);
		alertDialog.setTitle(R.string.account_inactive_title);

		if (isActive) {
			alertDialog.setMessage(getString(R.string.account_active_message));
		} else{
			alertDialog.setMessage(getString(R.string.account_inactive_message));
		}

		alertDialog.setPositiveButton("Ok", (dialog, which) -> login(null));
		alertDialog.show();
	}
}

File: mage/src/main/java/mil/nga/giat/mage/login/SignupActivity.kt
package mil.nga.giat.mage.login

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Patterns
import android.view.KeyEvent
import android.view.View
import android.view.inputmethod.InputMethodManager
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.core.graphics.ColorUtils
import androidx.lifecycle.ViewModelProvider
import dagger.hilt.android.AndroidEntryPoint
import mil.nga.giat.mage.R
import mil.nga.giat.mage.databinding.ActivitySignupBinding
import mil.nga.giat.mage.login.SignupViewModel.Account
import mil.nga.giat.mage.login.SignupViewModel.CaptchaState
import mil.nga.giat.mage.login.SignupViewModel.SignupError
import mil.nga.giat.mage.login.SignupViewModel.SignupState
import mil.nga.giat.mage.login.SignupViewModel.SignupStatus

@AndroidEntryPoint
open class SignupActivity : AppCompatActivity() {

   protected lateinit var binding: ActivitySignupBinding
   protected lateinit var viewModel: SignupViewModel

   public override fun onCreate(savedInstanceState: Bundle?) {
      super.onCreate(savedInstanceState)

      binding = ActivitySignupBinding.inflate(layoutInflater)
      setContentView(binding.root)

      binding.signupButton.setOnClickListener { signup() }
      binding.cancelButton.setOnClickListener { cancel() }
      binding.refreshCaptcha.setOnClickListener { getCaptcha() }

      binding.signupUsername.setOnFocusChangeListener { _: View, hasFocus: Boolean ->
         if (!hasFocus) {
            getCaptcha()
         }
      }

      binding.signupPassword.addTextChangedListener(object : TextWatcher {
         override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
         override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {}
         override fun afterTextChanged(s: Editable) {
            val sanitizedPasswordInputs: MutableList<String> = ArrayList()
            viewModel.account?.let {
               sanitizedPasswordInputs.add(it.username)
               sanitizedPasswordInputs.add(it.displayName)
               sanitizedPasswordInputs.add(it.email)
               sanitizedPasswordInputs.removeAll(setOf<Any?>(null))
            }

            val passwordStrengthFragment = supportFragmentManager.findFragmentById(R.id.password_strength_fragment) as PasswordStrengthFragment
            passwordStrengthFragment.setSanitizedList(sanitizedPasswordInputs)
            passwordStrengthFragment.onPasswordChanged(s.toString())
         }
      })

      binding.captchaText.setOnKeyListener { _, keyCode, _ ->
         if (keyCode == KeyEvent.KEYCODE_ENTER) {
            signup()
            return@setOnKeyListener true
         } else {
            return@setOnKeyListener false
         }
      }

      viewModel = ViewModelProvider(this).get(SignupViewModel::class.java)
      viewModel.signupState.observe(this) { state: SignupState -> onSignupState(state) }
      viewModel.signupStatus.observe(this) { status: SignupStatus? -> onSignup(status) }
      viewModel.captcha.observe(this) { captcha: String? -> onCaptcha(captcha) }
      viewModel.captchaState.observe(this) { state: CaptchaState -> onCaptchaState(state) }
   }

   protected fun onSignupState(state: SignupState) {
      if (state == SignupState.CANCEL) {
         done()
      } else {
         toggleMask(state === SignupState.LOADING)
      }
   }

   private fun onCaptchaState(state: CaptchaState) {
      binding.captchaProgress.visibility = if (state == CaptchaState.LOADING) View.VISIBLE else View.GONE
   }

   private fun onCaptcha(captcha: String?) {
      if (captcha != null) {
         binding.webView.visibility = View.VISIBLE
         binding.refreshCaptcha.visibility = View.VISIBLE
         val html = "<html><body style=\"background-color: ${backgroundColor()};\"><div style=\"width: 100%; height:100%;\" ><img src=\"$captcha\" width=\"100%\" height=\"100%\"\"/></div></body></html>"
         binding.webView.loadDataWithBaseURL(null, html, "text/html; charset=utf-8", "utf8", null)
      } else {
         binding.webView.visibility = View.GONE
         binding.refreshCaptcha.visibility = View.VISIBLE
      }
   }

   protected fun onSignup(status: SignupStatus?) {
      if (status == null) return

      if (status.success) {
         val isActive = status.user!!.get("active").asBoolean
         showSignupSuccessDialog(isActive)
      } else {
         when (status.error) {
            SignupError.INVALID_USERNAME -> {
               binding.captchaText.setText("")
               binding.usernameLayout.error = "Username not available"
            }
            SignupError.INVALID_CAPTCHA -> {
               binding.captchaTextLayout.error = "Invalid Captcha"
            }
            else -> {
               AlertDialog.Builder(this)
                  .setTitle("Signup Failed")
                  .setMessage(status.errorMessage)
                  .setPositiveButton(android.R.string.ok, null)
                  .show()
            }
         }
      }
   }

   protected open fun signup() {
      binding.displaynameLayout.error = null
      binding.usernameLayout.error = null
      binding.emailLayout.error = null
      binding.passwordLayout.error = null
      binding.confirmpasswordLayout.error = null

      val displayName: String = binding.signupDisplayname.text.toString()
      val username: String = binding.signupUsername.text.toString()
      val email: String = binding.signupEmail.text.toString()
      val phone: String = binding.signupPhone.text.toString()

      // are the inputs valid?
      if (displayName.isEmpty()) {
         binding.displaynameLayout.error = "Display name can not be blank"
         return
      }
      if (username.isEmpty()) {
         binding.usernameLayout.error = "Username can not be blank"
         return
      }

      if (email.isNotEmpty() && !Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
         binding.emailLayout.error = "Please enter a valid email address"
         return
      }

      val password: String = binding.signupPassword.text.toString()
      val confirmpassword: String = binding.signupConfirmpassword.text.toString()
      if (password.isEmpty()) {
         binding.passwordLayout.error = "Password can not be blank"
         return
      }

      if (confirmpassword.isEmpty()) {
         binding.confirmpasswordLayout.error = "Enter password again"
         return
      }

      if (password != confirmpassword) {
         binding.passwordLayout.error = "Passwords do not match"
         binding.confirmpasswordLayout.error = "Passwords do not match"
         return
      }

      val captchaText = binding.captchaText.text.toString()
      if (captchaText.isEmpty()) {
         binding.captchaTextLayout.error = "Captcha text cannot be blank"
         return
      }

      hideKeyboard()

      viewModel.signup(Account(username, displayName, email, phone, password), captchaText)
   }

   private fun getCaptcha() {
      viewModel.getCaptcha(binding.signupUsername.text.toString(), backgroundColor())
   }

   private fun cancel() {
      viewModel.cancel()
   }

   protected fun hideKeyboard() {
      val view: View = findViewById(android.R.id.content)
      val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
      imm.hideSoftInputFromWindow(view.windowToken, 0)
   }

   private fun showSignupSuccessDialog(isActive: Boolean) {
      val alertDialog = AlertDialog.Builder(this)
      alertDialog.setTitle(R.string.account_inactive_title)
      if (isActive) {
         alertDialog.setMessage(getString(R.string.account_active_message))
      } else {
         alertDialog.setMessage(getString(R.string.account_inactive_message))
      }
      alertDialog.setPositiveButton("Ok") { _: DialogInterface?, _: Int -> done() }
      alertDialog.show()
   }

   private fun toggleMask(visible: Boolean) {
      binding.progress.visibility = if (visible) View.VISIBLE else View.GONE
   }

   private fun backgroundColor(): String {
      val color = ColorUtils.compositeColors(ContextCompat.getColor(this, R.color.background), Color.WHITE)
      return String.format("#%06X", 0xFFFFFF and color)
   }

   private fun done() {
      startActivity(Intent(applicationContext, LoginActivity::class.java))
      finish()
   }
}